Map code review

PJBoy opened the pull request:
This is a dummy pull request used purely to aide code review

moehr-moehr committed:
first working version - may have graphics bug

first working version
may need a fix for graphics bug where some item rooms occasionally are missing a column of tiles

moehr-moehr committed:
Merge pull request #1 from moehr-moehr/m2maps

first working version - may have graphics bug

moehr-moehr committed:
deletes unnecessary backup of src

first working version
may need a fix for graphics bug where some item rooms occasionally are missing a column of tiles

moehr-moehr committed:
Merge pull request #2 from moehr-moehr/m2maps

deletes unnecessary backup of src

moehr-moehr committed:
conflict resolution

warning: Overwrote a non-zero byte in the header checksum
warning: Overwrote a non-zero byte in the global checksum

moehr-moehr committed:
Merge pull request #3 from moehr-moehr/m2maps

conflict resolution

moehr-moehr committed:
updates and wip readme

moehr-moehr committed:
Merge pull request #4 from moehr-moehr/m2maps

updates and wip readme

moehr-moehr committed:
wip working towards compliance

begins the process of bringing code into style compliance

moehr-moehr committed:
Merge pull request #5 from moehr-moehr/m2maps

wip working towards compliance

moehr-moehr committed:
bugfix: white flash when screen disabled

applies a bugfix that feels wrong but works, does not call disableLCD but rather just writes the bits to turn it off and then turns it back on without any of the rest of the routine.

moehr-moehr committed:
Merge pull request #6 from moehr-moehr/m2maps

bugfix: white flash when screen disabled

moehr-moehr committed:
continues working toward compliance

mostly updates to constants, and using them in the tables instead of raw values

moehr-moehr committed:
Merge pull request #7 from moehr-moehr/m2maps

continues working toward compliance

moehr-moehr committed:
fixed message box flash on item pickup

correctly disables screen during pickup, perhaps!

moehr-moehr committed:
Merge pull request #8 from moehr-moehr/m2maps

fixed message box flash on item pickup

PJBoy commented on SRC/bank_000.asm:
```diff
@@ -132,10 +132,20 @@ VBlankHandler: ;{ 00:0154
+			cp a, gameMode_paused
+			jr nz, .skipUpdateHudPaused
+				ld a, BANK(VBlank_updateStatusBarPaused)
+				ld [rMBC_BANK_REG], a
```
It looks as though you're missing the call to `VBlank_updateStatusBarPaused`

moehr-moehr replied:
moved to bank 10 - I moved the label as well since I didn't know if other stuff in the vanilla game needed the original label or jumped there. All of the status bar content has changed except the metroid counter. I could probably remove the vanilla label though if that's sensible, and put it back if that breaks things.


Edit: looking through it again, the conditional here just skips the normal update here if the game is paused, and in the handler that checks if the game is paused, does the pause update there. I'm not using this label to call the routine in this bank though, and have both the label I am using and the -paused label in bank 10. I think I'm going to refer to it by the name pauseAdjustSpritemap like I do further down, just cause I'm not sure it's happening during vBlank.

I'm not attached to any names I came up with so I'll happily change those to any suggested

PJBoy replied:
I didn't mean that you're missing `VBlank_updateStatusBarPaused` the routine, I meant you're setting the bank as if you're going to do a `call VBlank_updateStatusBarPaused`, but there is no `call` :O

moehr-moehr replied:
> I didn't mean that you're missing `VBlank_updateStatusBarPaused` the routine, I meant you're setting the bank as if you're going to do a `call VBlank_updateStatusBarPaused`, but there is no `call` :O

I figured it out

The setup here preps this later call to run OAM_DMA:
```
    .else_B:
            ; m2maps: skip updating gameplay status bar if game is paused
                ld a, [gameMode]
                cp a, gameMode_paused
                jr nz, .skipUpdateHudPaused
                    ld a, BANK(m2maps_pauseAdjustSpriteSetup)
                    ld [rMBC_BANK_REG], a
                    jr .endIf_B
                .skipUpdateHudPaused:			
            ; end m2maps block
        ld a, BANK(VBlank_updateStatusBar)
        ld [rMBC_BANK_REG], a
        call VBlank_updateStatusBar
    .endIf_B:
		
; End vblank
    call OAM_DMA ; Sprite DMA
    ld a, [bankRegMirror]
    ld [rMBC_BANK_REG], a
    ld a, $01
    ldh [hVBlankDoneFlag], a
    pop hl
    pop de
    pop bc
    pop af
reti
```

I tried all configs of commenting this m2maps block out and basically if it's totally removed, you get the normal HUD overwriting the pauseHUD sprites that are needed to show in-game-time, current items out of total items, and the metroid remaining sprite.

That routine to display the map HUD is called in a different block that handles pausing. I think originally I tried to call it from here but I had to put the call elsewhere but the setup over here.

PJBoy replied:
I can understand jumping past the call to `VBlank_updateStatusBar` via the `jr .endIf_B`, but is the `ld a, BANK(m2maps_pauseAdjustSpriteSetup) : ld [rMBC_BANK_REG], a` actually doing anything?

moehr-moehr replied:
Was able to finally get this down to just the jump logic! And remove those placeholder NOPs we mentioned on discord.

Source of the bug was the title screen load I had modded to load a map at the title screen, thinking it was needed to have one available at start of gameplay at the landing site.

What actually happened was, lacking a sane currentLevelBank value, the game loaded gargabe data the size of a map to the window, overwrote the stack pointer as you described, and then returned to a location that accidentally worked.

Then, the door warp to the landing site reloaded the map (which I didn't realize was happening), making the initial map load of garbage data appear like it had functioned correctly.

After ripping out the initial map code in bank 5 I was able to remove the placeholder NOPs and have the game run again. I still have to step through it to verify now that I have a handle on the BGB debugger (which is actually useful as opposed to whatever VBA one I was using prior).

Thanks so much!!!

PJBoy replied:
Well diagnosed, any solution that removes code is a good solution :D

PJBoy commented on SRC/bank_000.asm:
```diff
@@ -667,6 +681,13 @@ gameMode_LoadB: ;{ 00:0464
+		callFar farLoadMapTiles
+		ld a, [currentLevelBank]
+		ld [bankRegMirror], a
+		ld [rMBC_BANK_REG], a
```
Should switch bank with `switchBankVar [currentlLevelBank]` (and in other places)

PJBoy commented on SRC/bank_000.asm:
```diff
@@ -9778,6 +9851,42 @@ handleItemPickup: ;{ 00:372F
+				ld h, d
+				ld l, e
+				ld e, $21
+				ld d, $9c
```
Lets use `HIGH(vramDest_itemText)` and `LOW(vramDest_itemText) + 1`

PJBoy commented on SRC/bank_003.asm:
```diff
@@ -1062,7 +1062,15 @@ queen_neckPatternPointers: ;{ 03:6C8E - Indexed by queen_neckPattern
 
 ; Initialize Queen AI
 queen_initialize: ;{ 03:6D4A
-    ; Clear the entire page
```
RIP this comment

PJBoy commented on SRC/bank_005.asm:
```diff
@@ -135,7 +135,6 @@ reti ;}
 ; called by gameMode_boot
 loadTitleScreen: ;{ 05:408F
     call title_loadGraphics
-    ; Load HUD
```
Suspect you didn't mean to move this comment

PJBoy commented on SRC/maps/doors.asm:
```diff
@@ -2878,9 +2879,21 @@ door1EF:
     LOAD_SPR gfx_metGamma
     END_DOOR
-
+;;;;m2maps: 2 added door sets, copies of earlier doors to make map transitions work
```
Why add these new doors instead of modifying the existing ones?

moehr-moehr replied:
This was strictly a matter of me not knowing yet if LAMP or ROSE need those pointers at specific ROM locations due to how they handle the data. I'll check with Conner and Liam and if that's not the case, I will happily modify the originals.

PJBoy replied:
OK, that's what I suspected, all good

PJBoy reviewed (commented):
Think you should change the tabs to spaces.
I think it might be a good idea to prefix all of your new labels with something (like `m2map_`, say). I think there's some value in knowing what's vanilla and what's new at a glance.

moehr-moehr committed:
Updates banks other than 10 except doorlist

Need to test doorlist modification with LAMP before modifying the original door script array

Otherwise should have all proposed edits incorporated so far.

moehr-moehr committed:
Merge pull request #9 from moehr-moehr/m2maps

Updates banks other than 10 except doorlist

moehr-moehr committed:
Delete frenemyData.asm

delete LAMP test file

moehr-moehr committed:
bank05-adjustment

added a 'hijack' so bytes could be preserved, allowing LAMP to load the door array, since that function depends on specific address loactions

moehr-moehr committed:
Merge pull request #10 from moehr-moehr/m2maps

Delete frenemyData.asm

moehr-moehr committed:
bughunting

working on getting rid of that one last stray NOP around bank 0 line 3783

moehr-moehr committed:
Merge pull request #11 from moehr-moehr/m2maps

bughunting

moehr-moehr committed:
Remove title map load

-unnecessary since map loads on warp into landing site
-having no currentLevelBank value, loaded garbage data and broke the stack
-able to clean up bank 5 and also some extraneous code in bank 0 as a result

moehr-moehr committed:
Merge pull request #12 from moehr-moehr/m2maps

Remove title map load

moehr-moehr committed:
bank 10 to bank 1f

pre-emptively move bank 10 map functions to bank 1f to get out of the way of the color patch

moehr-moehr committed:
Merge pull request #13 from moehr-moehr/m2maps

bank 10 to bank 1f

PJBoy commented on SRC/bank_01f.asm:
```diff
@@ -0,0 +1,578 @@
+			ld b, a
+			ld a, [mapSamusLocatorYOffset]
+			add a, b
+			rl a
+			rl a
+			rl a
```
```suggestion
			sla a
			sla a
			sla a
```
Shift left makes more sense than rotate left when you don't need those high bits rotated into the low bits

PJBoy commented on SRC/bank_01f.asm:
```diff
@@ -0,0 +1,578 @@
+			ld b, a
+			ld a, [mapSamusLocatorXOffset]
+			add a, b
+			rl a
+			rl a
+			rl a
```
```suggestion
			sla a
			sla a
			sla a
```

PJBoy commented on SRC/bank_01f.asm:
```diff
@@ -0,0 +1,578 @@
+					ld [mapSamusLocatorXOffset], a
+	ret
+
+m2maps_pauseAdjustSpriteSetup:
```
This routine does a bunch of writes to VRAM and a bunch of writes to OAM. I think it would be sensible to do all of the VRAM writes first, then for OAM you can load $C000 (or `wram_oamBuffer`) in `de` and use `ld [de], a : inc de` everywhere, meaning you don't have to keep spelling out $C0nn addresses or even have the `mapDotsOAM` constants

PJBoy commented on SRC/bank_01f.asm:
```diff
@@ -0,0 +1,578 @@
+		;draw time hour hi tile
+			ld a, [gameTimeHours]
+			and $f0
+			sra a
+			sra a
+			sra a
+			sra a
```
```suggestion
			srl a
			srl a
			srl a
			srl a
```
Logical shift makes more sense for values that can't be negative

PJBoy commented on SRC/bank_01f.asm:
```diff
@@ -0,0 +1,578 @@
+			ld [$c003], a
+		;draw time hour hi tile
+			ld a, [gameTimeHours]
+			and $f0
```
No need for this `and`

PJBoy commented on SRC/bank_01f.asm:
```diff
@@ -0,0 +1,578 @@
+		;draw time minute hi tile
+			ld a, [gameTimeMinutes]
+			and $f0
+			sra a
+			sra a
+			sra a
+			sra a
```
```suggestion
			srl a
			srl a
			srl a
			srl a
```

PJBoy commented on SRC/bank_01f.asm:
```diff
@@ -0,0 +1,578 @@
+		;draw numbers for count of remaining items to find
+			ld a, [mapItemsFound]
+			and a, $F0
+			sra a
+			sra a
+			sra a
+			sra a
```
```suggestion
			srl a
			srl a
			srl a
			srl a
```

PJBoy commented on SRC/bank_01f.asm:
```diff
@@ -0,0 +1,578 @@
+			ld [$9c08], a
+		;draw numbers for count of remaining items to find
+			ld a, [mapItemsFound]
+			and a, $F0
```
No need for this `and`

PJBoy commented on SRC/bank_01f.asm:
```diff
@@ -0,0 +1,578 @@
+		;draw tile for max items to find tens and ones
+			ld a, total_items
+			and a, $f0
+			sra a
+			sra a
+			sra a
+			sra a
```
```suggestion
			srl a
			srl a
			srl a
			srl a
```

PJBoy commented on SRC/bank_01f.asm:
```diff
@@ -0,0 +1,578 @@
+			ld [$9c0b], a
+		;draw tile for max items to find tens and ones
+			ld a, total_items
+			and a, $f0
```
Ditto

PJBoy commented on SRC/bank_01f.asm:
```diff
@@ -0,0 +1,578 @@
+			ld [$9c2d], a
+			ld [$9c2e], a
+			ld [$9c2f], a
+			ld [$9c30], a
```
I would hazard a guess that this store to $9C30 isn't needed

PJBoy commented on SRC/bank_01f.asm:
```diff
@@ -0,0 +1,578 @@
+			ld h, mapIconsWram_hi
+			ld l, mapIconsWram_lo
+			;load the write address of sprite OAM to begin at
+			ld d, mapDotsOAM_hi
+			ld e, mapDotsOAM_lo
```
(See comment at start of routine)

PJBoy commented on SRC/bank_01f.asm:
```diff
@@ -0,0 +1,578 @@
+							inc hl
+							inc hl
+							inc hl
+							ld a, tile_blank
+							ld [de], a
+							inc de
+							ld [de], a
+							inc de
+							ld [de], a
+							inc de
+							ld [de], a
+							inc de
```
Don't think you should be doing this at all

PJBoy commented on SRC/bank_01f.asm:
```diff
@@ -0,0 +1,578 @@
+			.next:
+		;mask message area of pause window black
+			ld a, tile_blank
+			ld [$9c20], a
+			ld [$9c21], a
+			ld [$9c22], a
+			ld [$9c23], a
+			ld [$9c24], a
+			ld [$9c25], a
+			ld [$9c26], a
+			ld [$9c27], a
+			ld [$9c28], a
+			ld [$9c29], a
+			ld [$9c2a], a
+			ld [$9c2b], a
+			ld [$9c2c], a
+			ld [$9c2d], a
+			ld [$9c2e], a
+			ld [$9c2f], a
```
Probably goes without saying, but this would be better as a loop

PJBoy commented on SRC/bank_01f.asm:
```diff
@@ -0,0 +1,578 @@
+					;read byte at bc. If not FF do four inc HL
+						;read hl+ and write de+ four times			
+			;load the read address of window RAM
+			ld h, mapIconsWram_hi
+			ld l, mapIconsWram_lo
```
```suggestion
			ld h, HIGH(mapWram)
			ld l, LOW(mapWram)
```

PJBoy commented on SRC/bank_01f.asm:
```diff
@@ -0,0 +1,578 @@
+				ld a, set_item_collected
+				ld [de], a
+		.doNotClearItem:
+			;a bad implementation of converting hex to decimal for display
+			ld a, [mapItemsFound]
+			and $0f
+			cp $09
+			jr nz, .notTensOverNine
+				ld a, [mapItemsFound]
+				add a, $10
+				and $f0
+				jr .resume
+			.notTensOverNine:
+				ld a, [mapItemsFound]
+				inc a
+		.resume:
+			ld [mapItemsFound], a
```
```suggestion
    ld a, [mapItemsFound]
    add a, 1
    daa
    ld [mapItemsFound], a
```
If you do an add or sub on a BCD number, you can use `daa` to adjust the resulting value into the corresponding BCD number

PJBoy commented on SRC/bank_01f.asm:
```diff
@@ -0,0 +1,578 @@
+				ld d, a
+				ld [clearItemBank], a
+			;loop to set up low byte of de with offset for enemy index for SRAM low address
+				ld a, [currentLevelBank]
+				sub a, clearItem_bankOffset
+				inc a
+				and a, clearItem_loopLimit
+				ld e, a
+				;get a for loop once more
+				ld a, [clearItemIndex]
+				.loopCurrentCheck
+					dec e
+					jr z, .stopChecking
+					add a, clearItem_sramOffsetLo
+					jr .loopCurrentCheck
+				.stopChecking:
```
If `[currentLevelBank]` is Ch, then the loop runs a good 255 times (and then happens to churn out the correct result of `a = C0h`), looks like the `inc a` is supposed to follow the `and`.

PJBoy commented on SRC/bank_01f.asm:
```diff
@@ -0,0 +1,578 @@
+			jr nz, .notFirstBeam
+			jr .doNotClearItem
+		.gotItem:
+			;get sram bank offset from enemy wram
+				ld a, [clearItemDotLow]
+				add a, clearItem_enemyOffset
+				ld e, a
+				ld a, [clearItemDotHigh]
+				ld d, a
+				ld a, [de]
+				sub a, clearItem_sramOffsetLo
+				ld [clearItemIndex], a
+			;set up bank as high byte of de for SRAM address, is $bank - $9 + $c9
+				ld a, [currentLevelBank]
+				sub a, clearItem_bankOffset
+				sra a
+				sra a
+				add a, clearItem_sramOffsetHi
+				ld d, a
+				ld [clearItemBank], a
+			;loop to set up low byte of de with offset for enemy index for SRAM low address
+				ld a, [currentLevelBank]
+				sub a, clearItem_bankOffset
+				inc a
+				and a, clearItem_loopLimit
+				ld e, a
+				;get a for loop once more
+				ld a, [clearItemIndex]
+				.loopCurrentCheck
+					dec e
+					jr z, .stopChecking
+					add a, clearItem_sramOffsetLo
+					jr .loopCurrentCheck
+				.stopChecking:
+			;have the byte of SRAM buffer to update as 'item collected'
+				ld [clearItemIndex], a
+				ld e, a
+				ld a, set_item_collected
+				ld [de], a
```
```suggestion
        ; clearItemIndex = $C900 + ([room bank] - 9) * 40h + [enemy map ID] - 40h
        ;get sram bank offset from enemy wram
            ld a, [clearItemDotLow]
            add a, clearItem_enemyOffset
            ld e, a
            ld a, [clearItemDotHigh]
            ld d, a
            ld a, [de]
            sub a, clearItem_sramOffsetLo
            ld [clearItemIndex], a
        ;set up bank as high byte of de for SRAM address, is $bank - $9 + $c9
            ld a, [currentLevelBank]
            sub a, clearItem_bankOffset
            ld b, a
            ld c, 0
            srl b
            rr c
            srl b
            rr c
            ld a, b
            add a, clearItem_sramOffsetHi
            ld d, a
            ld [clearItemBank], a
        ;set up low byte of de with offset for enemy index for SRAM low address
            ld a, [clearItemIndex]
            add a, c
        ;have the byte of SRAM buffer to update as 'item collected'
            ld [clearItemIndex], a
            ld e, a
            ld a, set_item_collected
            ld [de], a
```
This version replaces the multiplication loop with a more conventional bitshift multiplication. It does `a * 40h` by (ref line 538) loading `a` into the high byte of `bc` and shifting `bc` right by 2, equivalent to `a * 100h / 4 = a * 40h`

PJBoy commented on SRC/bank_01f.asm:
```diff
@@ -0,0 +1,578 @@
+m2maps_farLoadMapTiles:
+		; Get screen index from coordinates
+		ldh a, [hSamusYScreen]
+		rl a
+		rl a
+		rl a
+		rl a
+		and $f0
```
```suggestion
		sla a
		sla a
		sla a
		sla a
```
No need for the `and` if using shift instead of rotate :)

PJBoy commented on SRC/bank_01f.asm:
```diff
@@ -0,0 +1,578 @@
+			jr nz, .down
+				ld a, [samusMapY]
+				sub a, $10
+				and a, $f0
```
`and` isn't needed

PJBoy commented on SRC/bank_01f.asm:
```diff
@@ -0,0 +1,578 @@
+			jr nz, .none
+				ld a, [samusMapY]
+				add a, $10
+				and a, $f0
```
Ditto

PJBoy commented on SRC/ram/wram.asm:
```diff
@@ -1471,7 +1474,26 @@ bombArray:: ;$DD30..5F: Bomb data. 10h byte slots
+;;;;m2maps: wram additions for m2maps data at $dd60
+loadNewMapFlag:: ds $01	;IN USE flag at endDoorScript in bank 0 to run load new map tiles
+mapLevelBankIndexOffset:: ds $01	;debugging vars, =(mapLevelBank-9)x2, used as an offset to adjust an index
+mapCollectionIndex:: ds $01	;IN USE $dd60, map index to load, 256 entries per level map, used in bank 10 to determine which map to load
```
```suggestion
mapCollectionIndex:: ds $01	;IN USE $dd60, map index to load, 256 entries per level map, used in bank 1f to determine which map to load
```

PJBoy commented on SRC/bank_01f.asm:
```diff
@@ -0,0 +1,578 @@
+	add hl, de
+		;now we have pointer to the proper map collection table to review (indexed by currentLevelBank)
+		;load the value at that submap table index to hl to find the proper submap
+		ld e, [hl]
+		inc hl
```
```suggestion
		ld e, [hl+]
```

PJBoy commented on SRC/bank_01f.asm:
```diff
@@ -0,0 +1,578 @@
+			ld [mapCollectionTableXY+1], a
+				;now we have pointer to the proper map (indexed by mapCollectionIndex)
+				;load it and prepare to read data from the map to Window VRAM targets
+				ld e, [hl]
```
It looks like this was supposed to be a `ld e, [hl+]`

PJBoy commented on SRC/bank_01f.asm:
```diff
@@ -0,0 +1,578 @@
+	add hl, de	
+	;hl now points to map pointer
+	;
+	ld e, [hl]
+	inc hl
```
```suggestion
	ld e, [hl+]
```

PJBoy commented on SRC/bank_01f.asm:
```diff
@@ -0,0 +1,578 @@
+        dec b
+		ld a, b
+		and $0f
+		cp a, $06
```
Maybe change 6 to `$10 - (map_width/2)` where `map_width = 20 ; Screen width (160px) / tile width 8px)

PJBoy commented on SRC/bank_01f.asm:
```diff
@@ -0,0 +1,578 @@
+			ld a, b
+			and $f0
+			ld b, a
+			inc de
+			inc de
+			inc de
+			inc de
+			inc de
+			inc de
+			inc de
+			inc de
+			inc de
+			inc de
+			inc de
+			inc de
```
```suggestion
			ld a, e
			add a, $0C ; (Or $20 - map_width, where $20 is the size of a tile)
			ld e, a
			ld a, d
			adc a, $00
			ld d, a
```

PJBoy commented on SRC/bank_01f.asm:
```diff
@@ -0,0 +1,578 @@
+	ld a, b
+    jr nz, :--
+	;note - hl and de should already be set up from before
+	ld d, mapIconsWram_hi
+	ld e, mapIconsWram_lo
```
```suggestion
	ld d, HIGH(mapWram)
	ld e, LOW(mapWram)
```

PJBoy commented on SRC/bank_01f.asm:
```diff
@@ -0,0 +1,578 @@
+					add a, $01
+					ld h, a
+					;load the value at that pointer to samus Y offset
+					ld e, [hl]
+					inc hl
```
```suggestion
					ld e, [hl+]
```

PJBoy commented on SRC/bank_01f.asm:
```diff
@@ -0,0 +1,578 @@
+					add a, $02
+					ld h, a
+					;load the value at that pointer to samus X offset
+					ld e, [hl]
+					inc hl
```
```suggestion
					ld e, [hl+]
```

PJBoy commented on SRC/bank_01f.asm:
```diff
@@ -0,0 +1,578 @@
+
+;below - here, we handle the items-found counter,
+;then we load the text for the appropriate item if it's not a refill
+m2maps_calcFoundEquipment:
```
Would be good to mention here that `b = [itemCollected]`
